src/ufuncs.c

```diff
@@ -1682,6 +1682,50 @@ static void relate_pattern_func(char** args, npy_intp* dimensions, npy_intp* ste
 }
 static PyUFuncGenericFunction relate_pattern_funcs[1] = {&relate_pattern_func};
 
+static char polygonize_dtypes[2] = {NPY_OBJECT, NPY_OBJECT};
+static void polygonize_func(char** args, npy_intp* dimensions, npy_intp* steps,
+                            void* data) {
+  GEOSGeometry* geom = NULL;
+  unsigned int n_geoms;
+
+  GEOS_INIT;
+
+  GEOSGeometry** geoms = malloc(sizeof(void*) * dimensions[1]);
+  if (geoms == NULL) {
+    errstate = PGERR_NO_MALLOC;
+    goto finish;
+  }
+
+  SINGLE_COREDIM_LOOP_OUTER {
+    n_geoms = 0;
+    SINGLE_COREDIM_LOOP_INNER {
+      if (!get_geom(*(GeometryObject**)cp1, &geom)) {
+        errstate = PGERR_NOT_A_GEOMETRY;
+        goto finish;
+      }
+      if (geom == NULL) {
+        continue;
+      }
+      geoms[n_geoms] = geom;
+      n_geoms++;
+    }
+
+    GEOSGeometry* ret_ptr = GEOSPolygonize_r(ctx, geoms, n_geoms);
+    if (ret_ptr == NULL) {
+      errstate = PGERR_GEOS_EXCEPTION;
+      goto finish;
+    }
+    OUTPUT_Y;
+  }
+
+finish:
+  if (geoms != NULL) {
+    free(geoms);
```

Comment:
In case of an error, the already created geometries need to be destroyed. You now deallocate the array holding the geometry pointers, but the geometries itself will form a memory leak if the array was partially filled already.

I think such a thing is already done in the Y_Y functions.

Reply:
Is it possible that this (`destroy_geom_arr(ctx, geom_arr, i - 1);`) is only needed for functions that release the GIL where we store the output in a temporary C array `geom_arr` which we only in the end transfer to a numpy array. 

While for the non-threaded versions, I suppose we directly write in the output object-dtype array, and numpy takes care of cleaning up this array? 
(in any case, I don't see such a manual clean-up in eg the geometry creation functions (linestrings, linearrings, polygons_with_holes, etc)

Reply:
You're right, this is not necessary. All geometries in `geoms` are borrowed from Python because they come directly from `get_geom`